Zend/zend_gc.c

```diff
@@ -245,6 +245,56 @@ static zend_gc_globals gc_globals;
 # define GC_TRACE(str)
 #endif
 
+
+#define GC_STACK_SEGMENT_SIZE (((4096 - ZEND_MM_OVERHEAD) / sizeof(void*)) - 2)
+
+typedef struct _gc_stack gc_stack;
+
+struct _gc_stack {
+	gc_stack        *prev;
+	gc_stack        *next;
+	zend_refcounted *data[GC_STACK_SEGMENT_SIZE];
+};
+
+#define GC_STACK_DCL(init) \
+	gc_stack *_stack = init; \
+	size_t    _top = 0;
+
+#define GC_STACK_PUSH(ref) \
+	do { \
+		if (UNEXPECTED(_top == GC_STACK_SEGMENT_SIZE)) { \
+			if (UNEXPECTED(!_stack->next)) { \
+				gc_stack *segment = emalloc(sizeof(gc_stack)); \
+				segment->prev = _stack; \
+				segment->next = NULL; \
+				_stack->next = segment; \
+			} \
+			_stack = _stack->next; \
+			_top = 0; \
+		} \
+		_stack->data[_top++] = (ref); \
+	} while (0)
+
+#define GC_STACK_POP() \
```

Comment:
So with this last change stack operations are done using macros.
IMO, it's a bit harder to read than functions.
I don't know that much about C or how exactly we are using it in PHP, but it it were a function wouldn't the compiler inline it anyway and we would have easier to read and change code?

Reply:
Agree. Functions would require passing arguments by pointer, but in case of inlining this probably won't make slowdown. I'll try.

Reply:
Done.